### PR TITLE
Dataframe constructor improvements

### LINKED ISSUE #993: (Python) DataFrame constructor improvements
I think the `DataFrame` constructor can be a lot more user friendly and consistent. Here how the constructor currently reacts to a bunch of different inputs (:heavy_check_mark:=works as expected, :x:=error, :large_orange_diamond:=works but has limitations)

:x: empty constructor
:heavy_check_mark: empty list
:heavy_check_mark: empty dict
:heavy_check_mark: dict of lists
:heavy_check_mark: dict of tuples
:large_orange_diamond: np.ndarray -> works but cannot specify columns
:x: np.ndarray with labeled columns
:x: list of lists
:heavy_check_mark: list of polars Series with unique names
:x: list of unnamed polars Series
:x: tuple of polars Series
:x: single polars Series
:heavy_check_mark: pandas DataFrame with named columns
:x: pandas DataFrame with unnamed columns
:large_orange_diamond: using `from_rows` on list of lists -> automatic column naming inconsistent with np.ndarray constructor; cannot specify `nullable`

_[See this notebook](https://github.com/stinodego/polars/blob/research/DataFrame_constructor.ipynb) for the actual error messages / results._

The way I see it, it shouldn't be difficult to support all these options with some proper control flow in the `__init__` of `DataFrame`. I'd suggest adding a `columns` keyword argument to help parsing stuff without column name info, and possibly a `dtype` argument, as it seems that there is no way to get around the automatic dtype inference of `polars` right now.

I do see a hurdle for the "list of unnamed polars Series", because the `name` attribute of the Series is set to an empty string instead of `None`. That makes it difficult to distinguish from the case with multiple identical names, which you do want to throw an error for. But that's a different subject.

If you agree with this direction, I'd be happy to implement this.

### PR DESCRIPTION
**Pull request by stinodego**
__This PR needs #1004 to be merged before the last test case will pass.__

Fixes #993 

Refer to the test cases to see exactly which cases are now implemented and what the corresponding behaviour is. The only one I have not implemented yet is np.ndarray with labeled columns; seems quite niche. Might take a look at that later. Otherwise, there's a whole bunch of new possibilities, and implemented quite elegantly too, if I say so myself.

I introduced a DeprecationWarning which catches and handles the case when `nullable` was defined as a positional argument, e.g. `DataFrame({'a': [1,2,3]}, False)`.

I did break one thing, which is the column names for the `ndarray` input; these were `0`, `1`, `2`, etc; which was inconsistent with the other automatic column naming of `column_0`, `column_1`, etc. As discussed we went with the latter option, but now the first one will change. I don't really know how to deprecate this prettily.

I will add a generous documentation docstring in a future pull request, together with Series.

### CODE REVIEW

**Comment by stinodego** [comment]
I'll rebase this later today and get everything ready. But your comments are welcome already :)

File: py-polars/tests/test_df.py [hunk 1]
```diff
@@ -16,17 +17,130 @@ def test_version():
...
+
+    # Single Series
+    df = pl.DataFrame(pl.Series("a", [1, 2, 3]))
+    truth = pl.DataFrame({"a": [1, 2, 3]})
+    assert df.frame_equal(truth)
+
+
+def test_init_seq_of_seq():
+    # List of lists
+    df = pl.DataFrame([[1, 2, 3], [4, 5, 6]], columns=["a", "b", "c"])
```
Comment (ritchie46) [inline py-polars/tests/test_df.py:93]: I think we should default to column orientation. That's the default way users should think when using polars. It is also a lot more performant.

It can be ambiguous in case of equal rows and columns:

```python
pl.DataFrame([[1, 2], [4, 5]], columns=["a", "b"])
```

is 

```
a, b
1, 4
2, 5
```

in column orientation and 


```
a, b
1, 2
4, 5
```

in row orientation

I think we should add another `kwarg` to dictate the orientation in case of the `list[list[]]`. That `kwarg` is only needed when we cannot infer the orientation from the shape.

Comment (stinodego) [reply py-polars/tests/test_df.py:93]: Having an extra `kwarg` for the orientation sounds like a great idea. This could also affect other calls like `ndarray`. I'll try implementing it.

**Comment by stinodego** [comment]
@ritchie46 Could you elaborate a little on the implementation of `from_rows` in the `PyDataFrame` class? Is there a reason it isn't used to parse the `np.ndarray` in the constructor (before my PR)? And why does it not take a `nullable` argument?

**Comment by ritchie46** [comment]
> @ritchie46 Could you elaborate a little on the implementation of from_rows in the PyDataFrame class? Is there a reason it isn't used to parse the np.ndarray in the constructor (before my PR)? And why does it not take a nullable argument?

Yes, the `from_rows` was created later for the edge case where you cannot create a Series from columnar data. This is the case in a database query for instance (most databases produce row-wise data).

The `from_rows` implementation is slower for multiple reasons:

* **row orientation** -> we write memory all over the place |  we have to use the builders/ or iterators with an extra layer of indirection
* **not typed** -> per value, we check the type (including null). In columnar data, we have the schema of the column and do that type dispatch check only once per column.

In the case of row wise data for `np.ndarray`, its probably faster to transpose the array and use column-wise constructors from there on.

The column-wise constructors have the `nullable` bool, so that we can dispatch creating a `Series` from:

* `&[Option<T>]`
* `&[T]`

Where `&[T]` is a fast path that has less branching.

So to summarize, we need to favor column-wise constructors and use `from_rows` only for certain edge cases.

**Comment by stinodego** [comment]
Okay, follow-up questions:
* Why is the `np.ndarray` always created as `nullable`, regardless of the setting in the dataframe constructor?
* I'm thinking about how to handle the list-of-lists input. Currently I'm using `from_rows`; do you see a better way?

**Comment by ritchie46** [comment]
>     Why is the np.ndarray always created as nullable, regardless of the setting in the dataframe constructor?

I believe this because the `ndarray` can have `NaN` values, which I infer as missing. But I am not sure of this anymore.

> I'm thinking about how to handle the list-of-lists input. Currently I'm using from_rows; do you see a better way?

In case of row orientation I would use `from_rows` indeed. In the other cases, I would pass every list to the `Series` constructor. This yields a `List[Series]`, thus a `DataFrame`

**Comment by stinodego** [comment]
Well, I'm leaving `nullable` in for now, as I feel like I don't know enough about it yet. If you feel like we should remove it, maybe we should make an issue for it. I added the orientation argument now; writing tests as we speak!

**Comment by stinodego** [comment]
The `orientation` argument defaulting to `column` is a pretty big 'break' though; construction of `np.ndarray` was done in row-orientation before. I can default it to `row` for `np.ndarray` only and add a deprecation warning?

**Comment by ritchie46** [comment]
> The orientation argument defaulting to column is a pretty big 'break' though; construction of np.ndarray was done in row-orientation before. I can default it to row for np.ndarray only and add a deprecation warning?

Seems good to me. I can create an issue with the **breaking** tag. Then we can accumulate the changes and apply them for the `0.9` release.

**Comment by stinodego** [comment]
Ok, I'm changing that. Then the PR should be good.

**Comment by stinodego** [comment]
All right, I think we're there. This definitely needs some documentation now, though. Will do some work on that tomorrow.

**Comment by ritchie46** [comment]
> I typed up a docstring. Not sure if that's the right format though. I couldn't find docstring conventions anywhere for polars.

Looks good. Its numpy style docstrings where the types may be omitted as they are already declared.
